README.md

A package provides an easy way to add facebook type multiple newsfeed imageview.

<p>
    <img src="https://github.com/EjazAmir/newsfeed_multiple_imageview/blob/master/screenshots/1.png?raw=true"/>
    <img src="https://github.com/EjazAmir/newsfeed_multiple_imageview/blob/master/screenshots/2.png?raw=true"/>
</p>

## Installation

You just need to add `newsfeed_multiple_imageview` as a [dependency in your pubspec.yaml file](https://flutter.io/using-packages/).

```yaml
dependencies:
  newsfeed_multiple_imageview: ^1.0.0
```

## How to use

```dart
import 'package:newsfeed_multiple_imageview/newsfeed_multiple_imageview.dart';

 List<String> _imageUrls = [];

          @override
          void initState() {
            super.initState();

            _imageUrls.add("https://media.istockphoto.com/photos/colorful-sunset-at-davis-lake-picture-id1184692500?k=20&m=1184692500&s=612x612&w=0&h=7noTRS8UjiAVKU92eIhPG17PIWVh-kCmH5jKX5GOcdQ=");
            _imageUrls.add("https://images.unsplash.com/photo-1573155993874-d5d48af862ba?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxzZWFyY2h8NXx8cGFya3xlbnwwfHwwfHw%3D&w=1000&q=80");
            _imageUrls.add("https://images.pexels.com/photos/158028/bellingrath-gardens-alabama-landscape-scenic-158028.jpeg?auto=compress&cs=tinysrgb&dpr=1&w=500");
            _imageUrls.add("https://images.pexels.com/photos/158028/bellingrath-gardens-alabama-landscape-scenic-158028.jpeg?auto=compress&cs=tinysrgb&dpr=1&w=500");
            _imageUrls.add("https://images.pexels.com/photos/158028/bellingrath-gardens-alabama-landscape-scenic-158028.jpeg?auto=compress&cs=tinysrgb&dpr=1&w=500");
          }

            NewsfeedMultipleImageView(
                imageUrls: _imageUrls,
              marginLeft: 10.0,
              marginRight: 10.0,
              marginBottom: 10.0,
              marginTop: 10.0,
            )
```


## Support

For support, email ejazaamir2013@gmail.com

Made with 💝


CHANGELOG.md
## 0.0.2

* Change in License file


lib/src/smart_image.dart
import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

import 'package:flutter/material.dart';
import 'package:transparent_image/transparent_image.dart';

class SmartImage extends StatelessWidget {
  final String src;
  final BoxFit? fit;
  final bool isPost;
  final double? radius;

  const SmartImage(this.src,
      {Key? key, this.fit, this.isPost = false, this.radius})
      : super(key: key);

  bool networkImage() => src.startsWith('https');
  //bool base64() => src.contains('[]');

  @override
  Widget build(BuildContext context) {
    return networkImage()
        ? FadeInImage.memoryNetwork(
            placeholder: kTransparentImage,
            image: src,
            fit: fit,
            imageErrorBuilder: (_, e, a) {
              return Container(
                alignment: Alignment.center,
                child: const Text(
                  "Error. Please check your internet connection",
                  style: TextStyle(
                    fontSize: 17,
                    fontWeight: FontWeight.bold,
                    fontFamily: 'Nunito',
                    color: Colors.black,
                  ),
                  textAlign: TextAlign.center,
                ),
              );
            },
          )
        : isPost
            ? Image.file(
                File(src),
                fit: fit,
              )
            : CircleAvatar(
                radius: radius,
                backgroundImage: MemoryImage(
                  imageDecoder(src),
                  //fit: fit,
                ),
              );
  }

  Uint8List imageDecoder(String image) {
    final List<int> list = List<int>.from(jsonDecode(image));
    return Uint8List.fromList(list);
  }
}


lib/src/multiple_image_view.dart
import 'package:flutter/material.dart';
import 'package:newsfeed_multiple_imageview/src/smart_image.dart';

class MultipleImageView extends StatelessWidget {
  final List<String> imageUrls;
  const MultipleImageView({
    Key? key,
    required this.imageUrls,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Row(
      mainAxisSize: MainAxisSize.max,
      mainAxisAlignment: MainAxisAlignment.center,
      crossAxisAlignment: CrossAxisAlignment.stretch,
      children: _contentRender(),
    );
  }

  List<Widget> _contentRender() {
    switch (imageUrls.length) {
      case 1:
        return _singleImageView();

      case 2:
        return _twoImageView();

      case 3:
        return _threeImageView();

      case 4:
        return _foureImageView();

      default:
        return _multipleImageView();
    }
  }

  List<Widget> _singleImageView() {
    return [
      Expanded(
        flex: 1,
        child: ClipRRect(
          borderRadius: BorderRadius.circular(20),
          child: SmartImage(
            imageUrls.first,
            fit: BoxFit.cover,
            isPost: true,
          ),
        ),
      ),
    ];
  }

  List<Widget> _twoImageView() {
    return [
      Expanded(
        flex: 1,
        child: ClipRRect(
          borderRadius: const BorderRadius.only(
            topLeft: Radius.circular(20),
            bottomLeft: Radius.circular(20),
          ),
          child: SmartImage(
            imageUrls.first,
            fit: BoxFit.cover,
            isPost: true,
          ),
        ),
      ),
      const SizedBox(width: 5),
      Expanded(
        flex: 1,
        child: ClipRRect(
          borderRadius: const BorderRadius.only(
            topRight: Radius.circular(20),
            bottomRight: Radius.circular(20),
          ),
          child: SmartImage(
            imageUrls.last,
            fit: BoxFit.cover,
            isPost: true,
          ),
        ),
      )
    ];
  }

  List<Widget> _threeImageView() {
    return [
      Expanded(
        flex: 1,
        child: ClipRRect(
          borderRadius: const BorderRadius.only(
            topLeft: Radius.circular(20),
            bottomLeft: Radius.circular(20),
          ),
          child: SmartImage(
            imageUrls[0],
            fit: BoxFit.cover,
            isPost: true,
          ),
        ),
      ),
      const SizedBox(width: 5),
      Expanded(
        flex: 1,
        child: Column(
          mainAxisSize: MainAxisSize.max,
          mainAxisAlignment: MainAxisAlignment.spaceBetween,
          crossAxisAlignment: CrossAxisAlignment.stretch,
          children: [
            Expanded(
              flex: 1,
              child: ClipRRect(
                borderRadius: const BorderRadius.only(
                  topRight: Radius.circular(20),
                ),
                child: SmartImage(
                  imageUrls[1],
                  fit: BoxFit.cover,
                  isPost: true,
                ),
              ),
            ),
            const SizedBox(height: 5),
            Expanded(
              flex: 1,
              child: ClipRRect(
                borderRadius:const BorderRadius.only(
                  bottomRight: Radius.circular(20),
                ),
                child: SmartImage(
                  imageUrls[2],
                  fit: BoxFit.cover,
                  isPost: true,
                ),
              ),
            ),
          ],
        ),
      ),
    ];
  }

  List<Widget> _foureImageView() {
    return [
      Expanded(
        flex: 2,
        child: ClipRRect(
          borderRadius: BorderRadius.only(
            topLeft: Radius.circular(20),
            bottomLeft: Radius.circular(20),
          ),
          child: SmartImage(
            imageUrls[0],
            fit: BoxFit.cover,
            isPost: true,
          ),
        ),
      ),
      const SizedBox(width: 5),
      Expanded(
        flex: 1,
        child: Column(
          mainAxisSize: MainAxisSize.max,
          mainAxisAlignment: MainAxisAlignment.spaceBetween,
          crossAxisAlignment: CrossAxisAlignment.stretch,
          children: [
            Expanded(
              flex: 1,
              child: ClipRRect(
                borderRadius:const  BorderRadius.only(
                  topRight: Radius.circular(20),
                ),
                child: SmartImage(
                  imageUrls[1],
                  fit: BoxFit.cover,
                  isPost: true,
                ),
              ),
            ),
            const SizedBox(height: 5),
            Expanded(
              flex: 1,
              child: ClipRRect(
                child: SmartImage(
                  imageUrls[2],
                  fit: BoxFit.cover,
                  isPost: true,
                ),
              ),
            ),
            const SizedBox(height: 5),
            Expanded(
              flex: 1,
              child: ClipRRect(
                borderRadius: const BorderRadius.only(
                  bottomRight: Radius.circular(20),
                ),
                child: SmartImage(
                  imageUrls[3],
                  fit: BoxFit.cover,
                  isPost: true,
                ),
              ),
            ),
          ],
        ),
      ),
    ];
  }

  List<Widget> _multipleImageView() {
    return [
      Expanded(
        flex: 2,
        child: ClipRRect(
          borderRadius: const BorderRadius.only(
            topLeft: Radius.circular(20),
            bottomLeft: Radius.circular(20),
          ),
          child: SmartImage(
            imageUrls[0],
            fit: BoxFit.cover,
            isPost: true,
          ),
        ),
      ),
      const SizedBox(width: 5),
      Expanded(
        flex: 1,
        child: Column(
          mainAxisSize: MainAxisSize.max,
          mainAxisAlignment: MainAxisAlignment.spaceBetween,
          crossAxisAlignment: CrossAxisAlignment.stretch,
          children: [
            Expanded(
              flex: 1,
              child: ClipRRect(
                borderRadius: const BorderRadius.only(
                  topRight: Radius.circular(20),
                ),
                child: SmartImage(
                  imageUrls[1],
                  fit: BoxFit.cover,
                  isPost: true,
                ),
              ),
            ),
           const SizedBox(height: 5),
            Expanded(
              flex: 1,
              child: ClipRRect(
                child: SmartImage(
                  imageUrls[2],
                  fit: BoxFit.cover,
                  isPost: true,
                ),
              ),
            ),
            const SizedBox(height: 5),
            Expanded(
              flex: 1,
              child: Stack(
                alignment: Alignment.center,
                fit: StackFit.expand,
                children: [
                  ClipRRect(
                    borderRadius: const BorderRadius.only(
                      bottomRight: Radius.circular(20),
                    ),
                    child: SmartImage(
                      imageUrls[3],
                      fit: BoxFit.cover,
                      isPost: true,
                    ),
                  ),
                  Positioned.fill(
                    child: Container(
                      alignment: Alignment.center,
                      decoration: BoxDecoration(
                        color: Colors.black.withOpacity(0.5),
                        borderRadius: const BorderRadius.only(
                          bottomRight: Radius.circular(20),
                        ),
                      ),
                      child: Text(
                        '+${imageUrls.length - 4}',
                        style: const TextStyle(
                          color: Colors.white,
                          fontSize: 22,
                          fontFamily: "Nulito",
                          fontWeight: FontWeight.w700,
                        ),
                      ),
                    ),
                  ),
                ],
              ),
            ),
          ],
        ),
      ),
    ];
  }
}


lib/src/newsfeed_multiple_imageview.dart
library newsfeed_multiple_imageview;

import 'package:flutter/material.dart';
import 'package:flutter_image_slideshow/flutter_image_slideshow.dart';
import 'package:newsfeed_multiple_imageview/src/smart_image.dart';

import 'package:newsfeed_multiple_imageview/src/multiple_image_view.dart';

class NewsfeedMultipleImageView extends StatelessWidget {

  final List<String> imageUrls;
  final double marginLeft;
  final double marginTop;
  final double marginRight;
  final double marginBottom;

  const NewsfeedMultipleImageView({
    Key? key,
    this.marginLeft = 0,
    this.marginTop = 0,
    this.marginRight = 0,
    this.marginBottom = 0,
    required this.imageUrls,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return LayoutBuilder(
      builder: (context, costraints) => Container(
        width: costraints.maxWidth,
        height: costraints.maxWidth,
        margin: EdgeInsets.fromLTRB(
          marginLeft,
          marginTop,
          marginRight,
          marginBottom,
        ),
        child: GestureDetector(
          child: MultipleImageView(imageUrls: imageUrls),
          onTap: () => Navigator.push(
            context,
            MaterialPageRoute(
              builder: (context) => ImageViewer(imageUrls: imageUrls),
            ),
          ),
        ),
      ),
    );
  }
}

class ImageViewer extends StatelessWidget {
  final List<String> imageUrls;
  const ImageViewer({Key? key,
    required this.imageUrls,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Scaffold(body: SafeArea(
      bottom: false,
      child: Container(
        // width: MediaQuery.of(context).size.width,
        // height: MediaQuery.of(context).size.height,
        color: Colors.black,
        child: SafeArea(
          top: false,
          left: false,
          right: false,
          child: Column(
            mainAxisAlignment: MainAxisAlignment.start,
            crossAxisAlignment: CrossAxisAlignment.start,
            mainAxisSize: MainAxisSize.max,
            children: [
              IconButton(
                onPressed: () => Navigator.pop(context),
                icon: const Icon(
                  Icons.close,
                  color: Colors.white,
                  size: 30,
                ),
              ),
              Expanded(
                child: ImageSlideshow(
                  initialPage: 0,
                  indicatorColor: Colors.red,
                  indicatorBackgroundColor: Colors.grey,
                  isLoop: imageUrls.length > 1,
                  children: imageUrls
                      .map(
                        (e) => ClipRect(
                      child: SmartImage(
                        e,
                        fit: BoxFit.contain,
                        isPost: true,
                      ),
                    ),
                  )
                      .toList(),
                ),
              ),
            ],
          ),
        ),
      ),
    ));
  }
}




example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:newsfeed_multiple_imageview/newsfeed_multiple_imageview.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'Newsfeed Multiple ImagesView',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const DemoApp(),
    );
  }
}

class DemoApp extends StatefulWidget {
  const DemoApp({Key? key}) : super(key: key);

  @override
  State<DemoApp> createState() => _DemoAppState();
}

class _DemoAppState extends State<DemoApp> {

  List<String> _imageUrls = [];

  @override
  void initState() {
    super.initState();

    _imageUrls.add("https://media.istockphoto.com/photos/colorful-sunset-at-davis-lake-picture-id1184692500?k=20&m=1184692500&s=612x612&w=0&h=7noTRS8UjiAVKU92eIhPG17PIWVh-kCmH5jKX5GOcdQ=");
    _imageUrls.add("https://images.unsplash.com/photo-1573155993874-d5d48af862ba?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxzZWFyY2h8NXx8cGFya3xlbnwwfHwwfHw%3D&w=1000&q=80");
    _imageUrls.add("https://images.pexels.com/photos/158028/bellingrath-gardens-alabama-landscape-scenic-158028.jpeg?auto=compress&cs=tinysrgb&dpr=1&w=500");
    _imageUrls.add("https://images.pexels.com/photos/158028/bellingrath-gardens-alabama-landscape-scenic-158028.jpeg?auto=compress&cs=tinysrgb&dpr=1&w=500");
    _imageUrls.add("https://images.pexels.com/photos/158028/bellingrath-gardens-alabama-landscape-scenic-158028.jpeg?auto=compress&cs=tinysrgb&dpr=1&w=500");
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: SafeArea(
        child: NewsfeedMultipleImageView(
            imageUrls: _imageUrls,
          marginLeft: 10.0,
          marginRight: 10.0,
          marginBottom: 10.0,
          marginTop: 10.0,
        ),
      ),
    );
  }
}
